### ISSUE TITLE
Question / Comment: getting dimensions of an image as it appears in the PDF

### ISSUE THREAD

**Issue by jmac105** [pymupdf/PyMuPDF#743]
Hi, thanks very much for your work on this repo it's excellent. My use case for it is to extract the 'useful' images from a PDF whilst trying to avoid as many unimportant images as possible. 

A common issue I've come across is that an image (such as a logo) may appear quite small in the PDF itself, but when extracted the actual image file is large both in dimensions and bytes size which makes it almost impossible to identify as an unimportant image. I've started then looking into how I can find the size of the image as it looks on the page which has led me to the `getImageBbox` function. 

The majority of PDFs I have tested `getImageBbox` do not work well with it, and raise the `unsupported image item` error. From reading other issues on here I believe this is because the images are inside page X Objects, for which I have put in the workaround suggested here https://github.com/pymupdf/PyMuPDF/issues/699#issuecomment-714541268. It's not important to me where on the page the image is, all that matters in this case is the dimensions (area) of the image as it is on the page. 

Am I correct in using these methods for this, or is there another way?

**Comment by JorjMcKie** [comment]
> Am I correct in using these methods for this, or is there another way?

Yes, just go ahead doing this. The restriction will be removed with the next verson.

**Comment by jmac105** [comment]
Thanks very much. While using that workaround, is the position of the Rect relative to the page or the XObject? I'm wondering if as a further filtering logic I could try to detect if the image is in the top/bottom 15% of the page, effectively meaning it was a part of the header or footer?

**Comment by JorjMcKie** [comment]
> While using that workaround, is the position of the Rect relative to the page or the XObject?

No, the bbox is relative to the original page.
There is ``doc.getPageXObjectList(pno)`` which returns a similar list for xobjects of the page - this time already with the xobject's bbox (as a tuple). That list as well is recursive: xobjects may invoke yet other xobjects.
The returned bbox sometimes happens to be incorrect - apparently though for deeply stacked xobjects only. For this reason I once decided to not allow extraction for xobjects alltogether: if in error, then that happens in MuPDF code - not mine 😎. So I don't want to be bothered with those issues.

> detect if the image is in the top/bottom 15% of the page

Sure, why not? You have the ``page.rect``, you have the bbox ... But please make sure to only use the **_unrotated_** page: the bbox pertains to it. You can use `page.setRotation(0)` without problem.

As per "unimportant" images: have you thought of other checks, e.g. unicolor? You could also check if the same image **_xref_** is used on many / all pages: a highly significant indicator for a logo. Same is true for xobject xrefs ... logos often are SVG images embedded as xobjects.

**Comment by jmac105** [comment]
Great, thanks. 

As regards other checks I'm using or considering all of the following: image bytes size, image dimensions (size and ratio), image bbox dimensions, position on page, if the same xref appears on multiple pages, if the sha1 hash of an extracted image matches the hash of a previously extracted image. Color is not something I'd consider but worth looking into, would I use pillow for that? 

I'm finding that PDFs are a complex subject... the more filtering I add the more likely it is that I will start to miss out important images

**Comment by JorjMcKie** [comment]
> I'm finding that PDFs are a complex subject...

You can say that again! Typical phenomenon for something that old: lots of stuff had to be added without being able to remove old specs for backward compatibility.
And then it is just a spec - nothing in a PDF file takes care of ensuring a valid structure. Very much opposed to XPS, which has a much better architecture - but a much smaller feature set ...

> would I use pillow for that

Probably yes. I successfully also used the quotient of `len(pix.getPNGData()) / len(pix.samples) < threshold`. If below some value like 0.05 (5%), forget it. But you will have to see what works best for you.

**Comment by jmac105** [comment]
One more question if I may: I've come across some PDFs where pageImageList does not match up with what's actually on the page. For example a 2 page PDFs with different images on each page, however the result for pageImageList is the same for each of the pages and each image seems to be referenced by an XObject which is again the same on both pages. 

Is there a method I'm missing to work out which page the images actually appear on? 

Result of `getPageImageList`, which comes back the same for both pages
```
[
    (11, 40, 3508, 142, 8, 'DeviceRGB', '', 'ExpertPdf_ndholdmhhjojkpooaeadfcaminnajckm', 'DCTDecode', 7), 
    (12, 42, 2048, 1097, 8, 'DeviceRGB', '', 'ExpertPdf_pdiealandcbbkgfkplldniokohnaiccn', 'DCTDecode', 7), 
    (13, 44, 1772, 1327, 8, 'DeviceRGB', '', 'ExpertPdf_ffdpbjmeaenmkjdmpeanejcgjojoicko', 'DCTDecode', 7), 
    (14, 36, 2048, 1360, 8, 'DeviceRGB', '', 'ExpertPdf_lcnggpelomogkcfoomfnfgbniagepncj', 'DCTDecode', 7), 
    (15, 41, 1129, 820, 8, 'DeviceRGB', '', 'ExpertPdf_pffdmhfpojlkkhekbopaebekilojobnh', 'DCTDecode', 7), 
    (16, 29, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_feignpjkgbkpkhdhajbacllcoangakdd', 'DCTDecode', 7), 
    (17, 43, 2048, 1092, 8, 'DeviceRGB', '', 'ExpertPdf_cakmaaegfiplkcdmbmpbhfdapmfcadia', 'DCTDecode', 7), 
    (18, 37, 3508, 709, 8, 'DeviceRGB', '', 'ExpertPdf_jbegcakhmmoakegpocbgkoikpoiofffm', 'DCTDecode', 7), 
    (19, 31, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_kdeebmeebikckjolajjfanjklabeoiie', 'DCTDecode', 7), 
    (20, 28, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_dpjenhnepnamklolbiigffnaepjeabia', 'DCTDecode', 7), 
    (21, 35, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_ilddfmhephdikphpodnjelnbphegfgnp', 'DCTDecode', 7), 
    (22, 34, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_eglikohgfgpgkifpacpkjfcpjihojhok', 'DCTDecode', 7), 
    (23, 32, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_madmhjakhbjekfecpbhllnbchipbaoab', 'DCTDecode', 7), 
    (24, 33, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_liaokjhfafnckjheoednddefcfblojhp', 'DCTDecode', 7), 
    (25, 30, 120, 120, 8, 'DeviceRGB', '', 'ExpertPdf_phjhelckoodhkpabaheknphmdmbiadlb', 'DCTDecode', 7)
]
```

Result of `getPageXObjectList`, which comes back the same for both pages
```
[(7, 'ExpertPdf_flgbafpghpikkcofbpcihlonjbkidhjo', 0, (0.0, 0.0, 1123.0, 1587.0))]
```

**Comment by JorjMcKie** [comment]
> One more question if I may

More than that: you are very welcome to ask questions 😊.

This way of storing resource information for pages is a fairly common one. Important resources types  are fonts, images and xobject. Resources specification is therefore one of the "inheritable" information types. Another one is page size, yet another one rotation.
This approach saves storing the same information repeatedly for every page.
So much for the theory.
PyMuPDF documentation discusses this same topic, by the way, at more than one place.

What I am doing when creating the mentioned list, is just reporting whatever specification I encounter - i.e. I do not check, whether some page really uses every images or font or xobject mentioned in the resp. lists.

This implies that you must be prepared to handle this situation, too. There are at least two ways to do that:
1. `page.getImageBbox()` returns an infinite rectangle if the image is not actually used by that page. Check it and react accordingly.
2. Execute `page.cleanContents(sanitize=True)` before making that list. Among a lot of other things, this method creates a valid list of resources actually needed by this page. Other activities include consolidating and syntax cleaning of the `/Contents` objects of the page.

And of course that XObject at xref 7 may indeed be used on both pages - find out by cleaning the page.

**Comment by jmac105** [comment]
Thank you, that makes sense now. I think I had come across that reading somewhere but forgot about it...

I've tried both those approaches and strangely neither seemed to work out. getImageBbox returned all Rects as being finite, and although cleanContents(sanitize=True) did change the list it still contained duplicate xrefs. The XObjects they refer to are used by the page, and their reference is 0 (the page). 

I think I have managed to achieve it though by checking if `doc[pno].rect.contains(bbox)` which so far looks to be quite reliable
